File: backend/app/domains/admin/schemas/provider_observability.py
from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel, Field


class SloSuggestion(BaseModel):
    metric: str
    current_value: float
    suggested_threshold: float
    unit: str
    rationale: str


class ProviderHealthCard(BaseModel):
    provider_key: str
    total_events: int
    failed_events: int
    failure_rate: float | None
    timed_out_events: int
    timeout_rate: float | None
    fallback_events: int
    fallback_rate: float | None
    retry_events: int
    retry_rate: float | None
    avg_retry_count: float | None
    avg_latency_ms: float | None
    p95_latency_ms: float | None
    slo_suggestions: list[SloSuggestion] = Field(default_factory=list)
    telemetry_missing: bool


class ProviderObservabilityRange(BaseModel):
    from_date: date = Field(alias="from")
    to_date: date = Field(alias="to")

    model_config = {"populate_by_name": True}


class ProviderObservabilitySnapshot(BaseModel):
    organization_id: str
    range: ProviderObservabilityRange
    generated_at: datetime
    providers: list[ProviderHealthCard]
    telemetry_missing: bool

File: frontend/src/lib/api/provider-observability.ts
import { apiRequest } from "@/lib/api/request";

export type SloSuggestion = {
  metric: string;
  current_value: number;
  suggested_threshold: number;
  unit: string;
  rationale: string;
};

export type ProviderHealthCard = {
  provider_key: string;
  total_events: number;
  failed_events: number;
  failure_rate: number | null;
  timed_out_events: number;
  timeout_rate: number | null;
  fallback_events: number;
  fallback_rate: number | null;
  retry_events: number;
  retry_rate: number | null;
  avg_retry_count: number | null;
  avg_latency_ms: number | null;
  p95_latency_ms: number | null;
  slo_suggestions: SloSuggestion[];
  telemetry_missing: boolean;
};

export type ProviderObservabilityRange = {
  from: string;
  to: string;
};

export type ProviderObservabilitySnapshot = {
  organization_id: string;
  range: ProviderObservabilityRange;
  generated_at: string;
  providers: ProviderHealthCard[];
  telemetry_missing: boolean;
};

export type ProviderObservabilityQuery = {
  from?: string;
  to?: string;
};

export async function getProviderObservabilitySnapshot(
  query?: ProviderObservabilityQuery,
): Promise<ProviderObservabilitySnapshot> {
  const params = new URLSearchParams();
  if (query?.from) params.set("from", query.from);
  if (query?.to) params.set("to", query.to);
  const qs = params.toString();
  return apiRequest<ProviderObservabilitySnapshot>(
    `/admin/provider-observability${qs ? `?${qs}` : ""}`,
  );
}
